design/kanister-kopia-integration.md

```diff
@@ -0,0 +1,594 @@
+- The server also uses TLS certificates to secure incoming connections to it.
+- Kanister users can configure a repository server via a newly added Custom
+   Resource called `RepositoryServer` as described in the following section.
+
+#### Custom Resource Definition
+
+This design proposes a new Custom Resource Definition (CRD) named
+`RepositoryServer`, to represent a Kopia repository server.
+It is a namespace-scoped resource, owned and managed by the controller in the
+`kanister` namespace. As mentioned above, the CRD offers a set of parameters to
+configure a Kopia repository server.
+
+To limit the controller's RBAC scope, all the `RepositoryServer` resources
+are created in the `kanister` namespace.
+
+A sample `RepositoryServer` resource created to interact with the Kopia
+repository of workloads running in the `monitoring` namespace looks like this:
+
+```yaml
+apiVersion: cr.kanister.io/v1alpha1
+kind: RepositoryServer
+metadata:
+   name: repository-monitoring
+   namespace: kanister
+   labels:
+      repo.kanister.io/target-namespace: monitoring
+spec:
+   storage:
+      # required: storage location info
+      secretRef:
+         name: location
+         namespace: kanister
+      # required: creds to access the location
+      # optional when using location type file-store
+      credentialSecretRef:
+         name: loc-creds
+         namespace: kanister
+   repository:
+      # repository must be created manually before creating server CR
+      # required: path for the repository - will be relative sub path
+      # within the path prefix specified in the location
+      rootPath: /repo/monitoring/
+      # required: password to access the repository
+      passwordSecretRef:
+         name: repository-monitoring-password
+         namespace: kanister
+      # optional: if specified, these values will be used by the controller to
+      # override default username and hostname when connecting to the
+      # repository from the server.
+      # if not specified, the controller will use generated defaults
+      username: kanisterAdmin
+      hostname: monitoring
+   server:
+      # required: server admin details
+      adminSecretRef:
+         name: repository-monitoring-server-admin
+         namespace: kanister
+      # required: TLS certificate required for secure communication between the Kopia client and server
+      tlsSecretRef:
+         name: repository-monitoring-server-tls
+         namespace: kanister
+      # required: repository users list
+      userAccessSecretRef:
+         name: repository-monitoring-server-access
+         namespace: kanister
+      # required: selector for network policy required to enable cross-namespace access to the server
+      networkPolicy:
+         namespaceSelector:
+            matchLabels:
+               app: monitoring
+         podSelector:
+            matchLabels:
+               pod: kopia-client
+               app: monitoring
+status:
```

Comment:
@pavannd1 We also discussed about adding a new string field `isReady` in the status subresource that would convey the users if RepositoryServer is ready to use or not
1. Value would either be "True" if ready
2. Or would be "False" if an error occurred in the controller